feat: add github-actions tag support

RichiCoder1 opened the pull request:
Adds support for GitHub Action tag versions in GitHub Action Workflow Files.

Removes deprecated old-style workflow support.

Not include:
* Git Sha Support

Closes #5733

rarkins commented on lib/manager/github-actions/__snapshots__/extract.spec.ts.snap:
```diff
@@ -1,5 +1,44 @@
+  Object {
+    "currentValue": "v1.0.0",
+    "datasource": "github-tags",
+    "depName": "actions/bin/shellcheck",
```
I don't think this is valid tags - perhaps needs lookupName? Also github.com/actions/bin doesn't exist either?

RichiCoder1 replied:
Ha, that's fair. I'll update the examples to something more realistic and maybe add a path nesting limit.

rarkins replied:
if it's a/b/c then the a/b represents the repo while c is file or folder name? And if there's multiple actions within one repo then they share the same tags? Ie tags are per repo and not per action

RichiCoder1 replied:
Hmmm. While actions [technically supports repo subpaths](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-using-a-public-action-in-a-subdirectory), the recommendation is [to have a single action per repo](https://docs.github.com/en/free-pro-team@latest/actions/creating-actions/about-actions#choosing-a-location-for-your-action). At least for this first pass of getting this working, I wouldn't support the path'd version as the way most people would likely use that is via `SHAs` and not tags. Or would require some sort of tag prefixing support.

rarkins replied:
If it's technically possible then we should ideally support it. In this case the lookupName would be `a/b`

RichiCoder1 replied:
I actually did end up doing this. The latest version of the regex will extract `owner/repo` and ignore the path.

RichiCoder1 commented:
Went ahead and improved the regex. It'll tolerate and ignore sub paths, prefers loose versioning (which is more correct?), and fixes tag to be lookupName instead of depName, which appears to be what `github-tags` expects.

viceice commented on lib/manager/github-actions/extract.ts:
```diff
@@ -1,30 +1,50 @@
+      const { lookupName, currentValue } = tagMatch.groups;
+      if (looseVersioning.api.isValid(currentValue)) {
+        const dep = {
+          lookupName,
```
```suggestion
          depName,
```

viceice commented on lib/manager/github-actions/extract.ts:
```diff
@@ -1,30 +1,50 @@
+      continue; // eslint-disable-line no-continue
+    }
+
+    const tagMatch = /^^\s+-?\s+?uses: (?<lookupName>[a-zA-Z-_]+\/[a-zA-Z-_]+)(?<path>.*)?@(?<currentValue>.+?)\s*?$/.exec(
```
```suggestion
    const tagMatch = /^^\s+-?\s+?uses: (?<depName>[a-zA-Z-_]+\/[a-zA-Z-_]+)(?<path>.*)?@(?<currentValue>.+?)\s*?$/.exec(
```
`depName` should always be present, renovate will fallback to `depName` for lookup, when no `lookupName` is set. So `lookupName` can be obmitted, if it has same value as `depName`.

viceice replied:
I think numbers are allowed too for github org and repo names?

RichiCoder1 replied:
You're correct! Swapping for `\w-`

viceice commented on lib/manager/github-actions/extract.ts:
```diff
@@ -1,30 +1,50 @@
+      line
+    );
+    if (tagMatch) {
+      const { lookupName, currentValue } = tagMatch.groups;
```
```suggestion
      const { depName, currentValue } = tagMatch.groups;
```

viceice commented on lib/manager/github-actions/extract.ts:
```diff
@@ -1,30 +1,50 @@
+    const tagMatch = /^^\s+-?\s+?uses: (?<lookupName>[a-zA-Z-_]+\/[a-zA-Z-_]+)(?<path>.*)?@(?<currentValue>.+?)\s*?$/.exec(
+      line
+    );
+    if (tagMatch) {
```
```suggestion
    if (tagMatch?.groups) {
```
To be typescript null check future proof

RichiCoder1 replied:
👍🏻

viceice commented on lib/manager/github-actions/extract.ts:
```diff
@@ -1,30 +1,59 @@
+        };
+        logger.debug(
+          {
+            depName: dep.depName,
+            currentValue: dep.currentValue,
```
```suggestion
            dep,
```

RichiCoder1 replied:
All the other examples were abbreviated, so wasn't sure. Changed

viceice replied:
ah, i see now. some manager log the dep and some not. 🤔 

@rarkins Should we pefer to remove the logging or should we add those logging to all managers? Maybe log at trace level?

rarkins replied:
No, we ideally shouldn't have that in managers as it bloats the logs. We should convert them to trace

RichiCoder1 commented:
Looking at other managers, should I specify `depType`?

viceice commented on lib/manager/github-actions/extract.ts:
```diff
@@ -1,30 +1,52 @@
+          commitMessageTopic: '{{depName}}} action',
+          datasource: githubTagsDatasource.id,
+          versioning: looseVersioning.id,
+          rangeStrategy: 'pin',
```
I'm not sure if this changes something for `loose` versioning. 🤔

RichiCoder1 replied:
Stole this from another manager, can drop it if it does nothing here haha

viceice replied:
```suggestion
```

Ok, so lets remove it.

viceice commented on lib/manager/github-actions/extract.ts:
```diff
@@ -1,30 +1,52 @@
+          currentValue,
+          commitMessageTopic: '{{depName}}} action',
+          datasource: githubTagsDatasource.id,
+          versioning: looseVersioning.id,
```
Why `loose` instead of default `semver` versioning? 🤔

RichiCoder1 replied:
There's not really a restriction on what you tag a version with, so github actions can (and often are) technically not semver-compliant. I original had this as semver, but from what I could tell `actions/checkout@v2` would have failed as it's an abbreviated version. Unless I'm misunderstanding?

viceice replied:
if we use loose here, we would update `v0` to `v1.0.0`. I'm not sure if this is desired. 🤔 
also it can happen to update from `v1.0.0` to `v2`

viceice commented:
> Looking at other managers, should I specify `depType`?

I don't think so.

viceice commented on lib/manager/github-actions/extract.ts:
```diff
@@ -1,30 +1,52 @@
+      continue; // eslint-disable-line no-continue
+    }
+
+    const tagMatch = /^^\s+-?\s+?uses: (?<depName>[\w-]+\/[\w-]+)(?<path>.*)?@(?<currentValue>.+?)\s*?$/.exec(
```
```suggestion
    const tagMatch = /^\s+-?\s+?uses: (?<depName>[\w-]+\/[\w-]+)(?<path>.*)?@(?<currentValue>.+?)\s*?$/.exec(
```
typo? 🙃

RichiCoder1 commented:
Switched to `github-releases` as the datasource, as this is more in line with what people expect and the marketplace. Per feedback, switched to `docker` versioning.

rarkins commented on lib/manager/github-actions/extract.ts:
```diff
@@ -7,24 +8,42 @@ export function extractPackageFile(content: string): PackageFile | null {
+        datasource: githubReleasesDatasource.id,
+        versioning: dockerVersioning.id,
+      };
+      logger.debug(dep, 'GitHub Action inside GitHub Workflow');
```
```suggestion
      logger.trace(dep, 'GitHub Action inside GitHub Workflow');
```

rarkins commented on lib/manager/github-actions/extract.ts:
```diff
@@ -7,24 +8,42 @@ export function extractPackageFile(content: string): PackageFile | null {
+        depName,
+        currentValue,
+        commitMessageTopic: '{{depName}}} action',
+        datasource: githubReleasesDatasource.id,
```
What was the reason to change this? Can't someone specify a tag in a workflow and it works? If so then we should look up tags? If GitHub enforces that you must *release* the tag then this is correct, but otherwise we can end up with a scenario where someone specifies a valid tag but we don't support it because we only look up releases.

RichiCoder1 replied:
It's a weird one. For a version to show up the marketplace, it must first be released and published. However, you're right that technically any tag can be referenced still (and GitHub uses this to tag their latest release with both the vMajor and vMajor.Minor.Patch tags).

I'm fine either way. Both github-tags and github-releases are applicable, just leaning toward the more stringent behavior.

rarkins replied:
I think let's go with tags then, as otherwise I can already predict future issues of people asking why Renovate isn't upgrading their Actions :)

RichiCoder1 replied:
Switched back to tags :+1:

rarkins commented on lib/manager/github-actions/extract.ts:
```diff
@@ -7,24 +8,42 @@ export function extractPackageFile(content: string): PackageFile | null {
+        currentValue,
+        commitMessageTopic: '{{depName}}} action',
+        datasource: githubReleasesDatasource.id,
+        versioning: dockerVersioning.id,
```
Is this because of a need to support `-compatibility` suffixes?

RichiCoder1 replied:
Twas at @viceice's recommendation. Apparently docker is slightly more stringent than `loose` in ways besides it's support for `-compat` suffixes? One other big difference I noted is that docker versioning will reject commit shas, which while technically supported seems correct.

rarkins replied:
First question: in this first implementation are we supporting *tags* and not SHAs? Because if SHA support is to be added later then we can worry about it later too.

Second question: at which stage will we ignore SHAs? e.g. will this extract function skip over them, or will they be included in the list of extract deps?

RichiCoder1 replied:
> First question: in this first implementation are we supporting tags and not SHAs? Because if SHA support is to be added later then we can worry about it later too.

My vote is support SHAs in a follow up.

> Second question: at which stage will we ignore SHAs? e.g. will this extract function skip over them, or will they be included in the list of extract deps?

Is there a way that's more efficient? I don't know the answer to this as a `renovatebot` newbie.

rarkins replied:
So one possibility would be the extract function does its own "Does this look like a version?" check and if not then sets a `skipReason` in the dep (e.g. unsupported value). Another is that it passes it through but the lookup process does the check based on the versioning. e.g. semver or docker versioning (whichever we ultimately decide to use here) returns `false` for `isValid()`. We want to make sure that SHAs are handled (ignored) gracefully.

RichiCoder1 replied:
Sounds like `skipReason` might be a better ux?

viceice replied:
@rarkins it's to Support v0 to v1 tag updates. Loose could cause a mixup from v0 to v1.0.1 or  v0.0.1 to v1.

viceice replied:
Maybe we need a custom versioning for this later, so we can use the compatibility function of versioning to suggest the right update.

RichiCoder1 replied:
Switched to extract w/ skip reason. I def think a custom version scheme will be the way forward in the future, to handle the happy path (published tag releases) as well as the less common cases (unpublished tags, commits, or branches).

rarkins commented:
@RichiCoder1 have you tested this latest code against any "real" repos?

RichiCoder1 commented:
>  have you tested this latest code against any "real" repos?

I've been running a regex-ified version against two of my repos with no issues (using docker & github tags). Gotten some updates, no big issues so far. We have pretty minor, sane usage though.

rarkins commented:
Let's make sure we have tested it against:
- nested presets 
- short ones (eg v1)
- regular semver 
- SHAs

rarkins commented:
I'll merge this once the current release is done

dargmuesli commented:
Will Renovate now start to update actions tags automatically or do I have to configure it to do so?

viceice commented:
> Will Renovate now start to update actions tags automatically or do I have to configure it to do so?

Yes, but the whitesource app is not yet updated, will be done today. So you shouldn't have to do anything.

RichiCoder1 commented:
@rarkins 🥳 Glad I could help with this! Sorry I couldn't help more with testing
